File: String Matching Algorithms and Their Applications/Rabin-Karp Algorithm/Rabin-Karp-Algorithm.c
#include <stdio.h>
#include <string.h>

#define d 256
#define q 101

void rabin_karp(char text[], char pattern[]) {
    int n = strlen(text);
    int m = strlen(pattern);
    
    if (m > n) {
        printf("Pattern is longer than the text.\n");
        return;
    }

    int h = 1, p_hash = 0, t_hash = 0;
    for (int i = 0; i < m - 1; i++)
        h = (h * d) % q;

    for (int i = 0; i < m; i++) {
        p_hash = (d * p_hash + pattern[i]) % q;
        t_hash = (d * t_hash + text[i]) % q;
    }

    int match_found = 0;
    for (int i = 0; i <= n - m; i++) {
        if (p_hash == t_hash) {
            int j;
            for (j = 0; j < m; j++)
                if (text[i + j] != pattern[j])
                    break;

            if (j == m) {
                printf("Pattern found at index %d\n", i);
                match_found = 1;
            }
        }

        if (i < n - m) {
            t_hash = (d * (t_hash - text[i] * h) + text[i + m]) % q;
            if (t_hash < 0)
                t_hash += q;
        }
    }

    if (!match_found)
        printf("Pattern not found in the text.\n");
}

int main() {
    char text[100], pattern[100];

    printf("Enter the text: ");
    scanf("%s", text);
    printf("Enter the pattern: ");
    scanf("%s", pattern);

    rabin_karp(text, pattern);
    return 0;
}

File: String Matching Algorithms and Their Applications/Rabin-Karp Algorithm/Rabin-Karp-Algorithm.py
def rabin_karp(text, pattern, d=256, q=101):
    n, m = len(text), len(pattern)

    if m > n:
        print("Pattern is longer than the text.")
        return

    h = pow(d, m - 1) % q  # Precomputed d^(m-1) % q
    p_hash, t_hash = 0, 0 
    match_found = False

    # computing initial hash values
    for i in range(m):
        p_hash = (d * p_hash + ord(pattern[i])) % q
        t_hash = (d * t_hash + ord(text[i])) % q

    # Sliding over text
    for i in range(n - m + 1):
        if p_hash == t_hash and text[i:i + m] == pattern:
            print(f"Pattern found at index {i}")
            match_found = True

        # Rolling Hash
        if i < n - m:
            t_hash = ((t_hash - ord(text[i]) * h) * d + ord(text[i + m])) % q
            if t_hash < 0:
                t_hash += q

    if not match_found:
        print("Pattern not found in the text.")

if __name__ == "__main__":
    text = input("Enter the text: ").strip()
    pattern = input("Enter the pattern: ").strip()

    if text and pattern:
        rabin_karp(text, pattern)
    else:
        print("Pattern and text cannot be empty.")
